python/python/lance/arrow.py

```diff
@@ -11,14 +11,373 @@
 #  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
-from typing import Optional, Union
+from pathlib import Path
+from typing import Callable, Iterable, Optional, Union
 
 import pyarrow as pa
 
 from .lance import BFloat16
 from .lance import bfloat16_array as bfloat16_array
 
 
+class ImageURIType(pa.ExtensionType):
+    def __init__(self):
+        pa.ExtensionType.__init__(self, pa.string(), "lance.arrow.image_uri")
+
+    def __arrow_ext_serialize__(self):
+        return b""
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, storage_type, serialized):
+        return ImageURIType()
+
+    def __arrow_ext_class__(self):
+        return ImageURIArray
+
+    def __arrow_ext_scalar_class__(self):
+        return ImageURIScalar
+
+
+class EncodedImageType(pa.ExtensionType):
+    def __init__(self):
+        # TODO: use pa.BinaryView once available?
+        pa.ExtensionType.__init__(self, pa.binary(), "lance.arrow.encoded_image")
+
+    def __arrow_ext_serialize__(self):
+        return b""
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, storage_type, serialized):
+        return EncodedImageType()
+
+    def __arrow_ext_class__(self):
+        return EncodedImageArray
+
+    def __arrow_ext_scalar_class__(self):
+        return EncodedImageScalar
+
+
+class FixedShapeImageTensorType(pa.ExtensionType):
+    def __init__(self, arrow_type, shape):
+        pa.ExtensionType.__init__(
+            self,
+            pa.fixed_shape_tensor(arrow_type, shape),
+            "lance.arrow.fixed_shape_image_tensor",
+        )
+
+    def __arrow_ext_serialize__(self):
+        return self.storage_type.__arrow_ext_serialize__()
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, storage_type, serialized):
+        import json
+
+        deserialized = json.loads(serialized.decode())
+        return FixedShapeImageTensorType(storage_type.value_type, deserialized["shape"])
+
+    def __arrow_ext_class__(self):
+        return FixedShapeImageTensorArray
+
+    def __arrow_ext_scalar_class__(self):
+        return FixedShapeImageTensorScalar
+
+
+class ImageURIArray(pa.ExtensionArray):
```

Comment:
Should we have one base `ImageArray(pa.ExtensionArray)` and concrete class for each different storage type?

It is a lot of work to educate users which class to use / trade offs.

Reply:
Agreed!
I was going to do a separate PR for the `ImageArray`, but might as well do it here.

Reply:
Added ImageArray.